In [19]:
##Importing neccessary libraries


import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

In [20]:
"""
This function helps in abstracting the 
underlying method by which ppandas reads in datasets
This aids portability and function behaviour can be changed in a single place.

This function is also important cos I can rasie exceptions or run checks, 
that are not included in the pandas read function in one place.
"""




def data_loader(filename: str):

  """

  this function reads in a dataset, just by passing in the path/name of 
  the file

  params: filename- This is the name/path to the file with format.
  

  """
  

  dataframe = pd.read_csv(filename)
 

  return dataframe



In [21]:
##Loading the dataset to a dataframe



df = data_loader('fuel_file.csv')



In [37]:
##Checking out the general information on the dataset


def summary_stat(dataframe):
  print("<-------------- Summary Statistics------------>")

  print(df.info())
  print("--------------------------------------------")

  print(df.dtypes)

  print("--------------------------------------------")


In [38]:
summary_stat(df)


<-------------- Summary Statistics------------>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB
None
---------

From the above stat, it shows that there 29523 entries(rows) with 11 columns 5 of the columns are of the Float64 datatype, 2 are Ints and 4 are string objects. The file size on disk is ~2.5mb  




In [36]:
df.describe()

utility_id_ferc1  ...  fuel_cost_per_mmbtu
count      29523.000000  ...         29523.000000
mean         118.601836  ...            19.304354
std           74.178353  ...          2091.540939
min            1.000000  ...           -41.501000
25%           55.000000  ...             1.940000
50%          122.000000  ...             4.127000
75%          176.000000  ...             7.745000
max          514.000000  ...        359278.000000

[8 rows x 7 columns]

In [43]:
#Check for missing values
"""
A function will iterate through
the column names and check if there's a missing value
if there is , pprint the count,this will reduce the assumption
to think that just a column has missing values
as with this current dataset, in the real world, there might be more 
missing data in different columns
"""


False

In [56]:
for column in column_list:
  if df.isna.any().any() == True:
    columns_with_missing= list(df.loc[:,df.isnull().any()].columns)





False
False
False
False
False
True
False
False
False
False
False


In [67]:

df.isna().any().any()

True

In [74]:
columns_with_missing= list(df.loc[:,df.isnull().any()].columns)

In [75]:
columns_with_missing

['fuel_unit']